EIPS/eip-2253.md

```diff
@@ -0,0 +1,172 @@
+
+The current proposal will avoid duplication of effort - dApps will not need to ask the user to create another address book with them. And it also protects the user's privacy, while increasing usability.
+
+## Specification
+
+New JSON-RPC method: `wallet_getAddressBook`.
+
+**Arguments:**
+- filter object, optional:
+  - `count` - type `uint256`, the number of address book records requested; optional - when missing, the entire address book is requested
+  - `fields` - type `string[]`, address book record fields that are requested; optional - when missing, all fields are requested
+
+**Result:**
+- array with address book records
+  - `address` - type `address`, required
+  - `name` - type `string`, optional, may not be returned if user refuses
+  - `icon` - type `base64`, token icon; optional
+
+### Examples
+
+**1) A request to return the entire address book, with all fields:**
+```
+{
+  "id":1,
+  "jsonrpc": "2.0",
+  "method": "wallet_getAddressBook",
+  "params": []
+}
+```
+Result:
+
+```
+{
+  "id":1,
+  "jsonrpc": "2.0",
+  "result": [
+    {
+      "address": "0x0000000000000000000000000000000000000001",
+      "name": "My Friend",
+      "icon": "data:image/gif;base64,R0lGODlhAQABAIABAP///wAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
+    },
+    {
+      "address": "0x0000000000000000000000000000000000000002",
+      "name": "Beer Buddy"
+    }
+  ]
+}
+```
+
+**2) A request to return one address book item, with all fields:**
+```
+{
+  "id":1,
+  "jsonrpc": "2.0",
+  "method": "wallet_getAddressBook",
+  "params": [
+    {
+      "count": 1,
+    }
+  ]
+}
+```
+Result:
+
+```
+{
+  "id":1,
+  "jsonrpc": "2.0",
+  "result": [
+    {
+      "address": "0x0000000000000000000000000000000000000001",
+      "name": "My Friend",
+      "icon": "data:image/gif;base64,R0lGODlhAQABAIABAP///wAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
+    }
+  ]
+}
+```
+
+**3) A request to return three address book items, only containing the address:**
+```
+{
+  "id":1,
+  "jsonrpc": "2.0",
+  "method": "wallet_getAddressBook",
+  "params": [
+    {
+      "count": 3,
+      "fields": ["address"]
+    }
+  ]
+}
+```
+Result:
+
+```
+{
+  "id":1,
+  "jsonrpc": "2.0",
+  "result": [
+    {
+      "address": "0x0000000000000000000000000000000000000001"
+    },
+    {
+      "address": "0x0000000000000000000000000000000000000002"
+    },
+    {
+      "address": "0x0000000000000000000000000000000000000003"
+    }
+  ]
+}
+```
+
+
+### UI Best Practices
+
+The wallet should display a UI to the user, showing the request. The user can accept or reject the request.
+
+If the entire address book is requested, the total number of address book items will be shown. However, even in this case, the user should be able to make a selection and restrict exposed fields.
+
+If an address book selection is requested, the user will have to select the items from the address book list.
```

Comment:
Would it make sense to have a "reason" in the request? So that the backend can tell the user 

> A request has come in from xxx to see `1` account, with reason: "select a friend to invite"` 
> Approve?

Reply:
This was also [proposed](https://ethereum-magicians.org/t/eip-2253-add-wallet-getaddressbook-json-rpc-method/3592/6) by @danfinlay :
> One parameter that might go well with this (and any permission, actually) would be a justification field

At that point, [my answer](https://ethereum-magicians.org/t/eip-2253-add-wallet-getaddressbook-json-rpc-method/3592/9?u=loredanacirstea) was:
> I understand the justification field would be in the permissions proposal. In which case, I agree. If the `wallet_getAddressBook` is used by the dApp (outside the permissions system), it is probably triggered by a user’s action and the dApp UI should provide the justification.

But I rethought this and you both are right. Regardless of the dApp's UI, it is reassuring for the user to see that the `reason`/`justification` on the wallet matches what the dApp shows.

`reason`/`justification`/`message` seem good - not sure which has been more used in other EIPs, with the same meaning.

Reply:
`justification` was added.